### PR TITLE
[do not merge] crypto backend

### PR DESCRIPTION
**Pull request by samoht**
Just opening a PR to add comments more easily

### CODE REVIEW

File: lib/chunck/irmin_chunck.ml [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+
+	(*	       
+	let extract_size_data x = 
+	  (get_chunck_length x), (get_chunck_data x)
+	 *)			 
+	type chunck_type =
+	  | Indirect 
+	  | Data
+	      
+	let chunck_type_to_int = function
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: usually called: `type_of_int` so you can write: `Chunk.type_of_int`

File: lib/chunck/irmin_chunck.ml [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+	  | Indirect -> 0
+	  | Data -> 1		      
+		      
+	let int_to_chunck_type = function
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: usually called `int_of_type` so you can write `Chunk.int_of_type`

File: lib/chunck/irmin_chunck.ml [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+	  | _ -> failwith "Unknow type"	       
+			  
+	let create_indirection t chunck_len =
+	  let c = Cstruct.create t.size in
+	  Cstruct.memset c 0x00;
+	  set_chunck_type c (chunck_type_to_int Indirect);
+	  set_chunck_length c chunck_len;
+	  c
+	    
+	let create_chunck t chunck_len =
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: just `create`

File: lib/chunck/irmin_chunck.ml [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+	  set_chunck_length c chunck_len;
+	  c
+(*
+	let concat x = Cstruct.concat x 
+ *)
+      end
+
+
+    module Tree = struct
+	open Chunck
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: avoid opening modules if possible (it makes reading code harder)

File: lib/chunck/irmin_chunck.ml [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+	    (fun x ->
+	     match (int_to_chunck_type (get_chunck_type node)) with 
+	     | Data -> Lwt.return 0
+	     | Indirect -> (height t x) >>= (fun y -> Lwt.return (1 + y))
+	    )
+		  *)
+	      	    
+	let walk_to_list t root =
+	  let rec aux_walk t node =
+	    match (int_to_chunck_type (get_chunck_type node)) with
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: you can remove the external parenthesis:

```ocaml
match Chunk.type_of_if (Chunk.get_type node) with
...
```

File: lib/chunck/irmin_chunck.ml [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+     Where type define if the chunck contain data or indirection, size represent the data length to 
+     consider, and data field is the payload. 
+     *)
+	       
+    module Chunck = struct 
+	
+	let hash_length = K.length
+			    
+	let l_type = 1
+	let get_chunck_type x = Cstruct.get_uint8 x 0
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: every function here are related to chunks (are they are in the `Chunk` module). You don't need to repeat `chunk` in the function name.

File: lib/chunck/irmin_chunck.ml [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+	       let nbr_child = get_chunck_length node in 
+	       let rec loop i accu =
+		 if i < nbr_child then
+		   (get_child t node i) >>=
+		     (fun x ->
+		      aux_walk t x >>=
+			(fun y ->
+			 (loop (i+1) (List.append accu y))
+			)
+		     )
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: you can write that without all the parenthesis:

```ocaml
get_child t node i >>= fun x ->
aux_walk t x        >>= fun y ->
loop (i+1) (List.append accu y)
```

File: lib/chunck/irmin_chunck.ml [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+		 loop (i+1) rest
+	     in
+
+	     let calc = loop 0 l in
+	     
+	     AO.add t.db indir >>=
+	       (fun x ->
+		(aux_botom_up t calc (x::r))
+	       )
+
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: same here

```ocaml
AO.add t.db indir >>= fun x ->
aux_botom_up t calc (x::r)
```

File: lib_test/test_krypto_chunck.ml~ [hunk 9]
```diff
@@ -0,0 +1,47 @@
+(*
```
Comment (samoht) [inline lib_test/test_krypto_chunck.ml~]: this file should not be here `*~`

File: lib_test/test_chunck.ml~ [hunk 10]
```diff
@@ -0,0 +1,47 @@
+(*
```
Comment (samoht) [inline lib_test/test_chunck.ml~]: this file should not be here

**Comment by samoht** [comment]
any update?

File: lib/index/irmin_index.ml [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+module Log = Log.Make(struct let section = "AOI" end)
+
+
+module type PERSISTANT_INDEX = sig
+
+    type index
+    type key
+
+    val find: index -> key
```
Comment (samoht) [inline lib/index/irmin_index.ml]: would be better:
```
val find: index -> key option
```

**Comment by samoht** [comment]
ping?

File: Makefile [hunk 12]
```diff
@@ -1,15 +1,9 @@
-TESTS  ?= --enable-tests
```
Comment (samoht) [inline Makefile]: Please do not overwrite the non-generated Makefile. The previous Makefile added various hooks which are necessary.

File: _oasis [hunk 13]
```diff
@@ -82,6 +83,27 @@ Library "irmin-unix"
...
+  FindlibParent:    irmin
+  BuildDepends:     irmin, nocrypto
+  Modules:          Irmin_krypto, Irmin_krypto_cipher, Irmin_krypto_km
+
+Library "irmin-link"
+  Path:             lib/link
+  Findlibname:      link
+  FindlibParent:    irmin
+  BuildDepends:     irmin, irmin.mem, irmin.unix, irmin.fs
+  Modules:          Irmin_link
```
Comment (samoht) [inline _oasis]: One of the reason to split `Irmin_link` into the various other backends would be to avoid depending on `irmin.unix` here. Might be other ways to avoid this (for instance by having a functor of `Irmin.FS.IO`). In any case, we don't want to depend on `*.unix` to make it compile to a unikernel later.

File: lib/chunck/irmin_chunck.ml [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+	    let add_chunck () =
+	      AO.add t.db chunck >>= fun x ->
+	      Chunck.set_data (K.to_raw x) 0 indir offset_hash Chunck.hash_length;
+	      Lwt.return_unit
+	    in
+	    Lwt.join [add_chunck (); loop (offset_first_level + 1)]
+	in
+	loop 0 >>= (fun _ ->
+		    AO.add t.db indir >>=
+		      (fun x ->
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: you can remove the `(fun x -> Lwt.return x)`

File: lib/chunck/irmin_chunck.ml [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+		   )
+      in
+      
+      let main_loop () =
+	let rec first_level i =
+	  match i with
+	  | e when e == nbr_indir_bloc -> 
+	      Lwt.return []
+	  | e when e < nbr_indir_bloc ->
+	     (split_data_first_level i) >>=
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: you can remove almost all the parenthesis in that block. The only needed one are the ones around `(i+1)` and `(x::y)`.

Also you can avoid the somehow verbose `>>= fun y -> Lwt.return (x::y)` and use `>|= fun y -> x::y`

File: lib/chunck/irmin_chunck.ml [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+      in
+      (main_loop ()) >>= Tree.bottom_up t
+
+
+
+    let mem t key =
+      try
+	AO.mem t.db key
+      with
+      | Not_found -> return_false
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: `AO.mem` should not raise an exception.

File: lib/chunck/irmin_chunck.ml [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+
+    let mem t key =
+      try
+	AO.mem t.db key
+      with
+      | Not_found -> return_false 
+
+
+    let iter _t (_fn : key -> value Lwt.t -> unit Lwt.t) =
+      failwith "TODO"
```
Comment (samoht) [inline lib/chunck/irmin_chunck.ml]: how do the tests pass if that function is not implemented? :-)

File: lib/chunck/irmin_chunck.mli [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+(** Chunck Module 
+
+     All functions for manipulate chuncks representation, a chunck is represented as above : 
+      --------------------------
+     | uint8_t type             | 
+     ---------------------------
+     | uint16_t chunck_length   |
```
Comment (samoht) [inline lib/chunck/irmin_chunck.mli]: what is the `chunk_length`? why is this necessary instead of using `Cstruct.len`?

File: lib/ir_hash.ml [hunk 19]
```diff
@@ -19,13 +19,14 @@ open Sexplib.Std
...
-  val of_raw: Cstruct.t -> t
-end
-
+    include Ir_hum.S	      
+    val digest: Cstruct.t -> t
+    val has_kind: [> `SHA1] -> bool
+    val to_raw: t -> Cstruct.t
+    val of_raw: Cstruct.t -> t
+    val length:int
+  end
```
Comment (samoht) [inline lib/ir_hash.ml]: why the extra spaces?

File: lib/ir_s.ml [hunk 20]
```diff
@@ -15,28 +15,28 @@
...
+	val config: t -> Ir_conf.t
+	val contents_t: t -> Contents.t
+	val node_t: t -> Node.t
+	val commit_t: t -> Commit.t
+	val tag_t: t -> Tag.t
+	val read_node: t -> key -> Node.key option Lwt.t
+	val mem_node: t -> key -> bool Lwt.t
+	val update_node: t -> key -> Node.key -> unit Lwt.t
+      end
   end
```
Comment (samoht) [inline lib/ir_s.ml]: same here, why the extra spaces? this makes the patch looks bigger than it is.

File: lib/ir_security.ml [hunk 21]
```diff
@@ -0,0 +1,100 @@
+
```
Comment (samoht) [inline lib/ir_security.ml]: `ir_security.ml` and `ir_security.mli` are not used anywhere ... are you using them? if that's the case would be better to move this in the `krypto/` directory.

File: lib/irmin.mli [hunk 22]
```diff
@@ -379,7 +379,22 @@ module type AO = sig
...
+
+(** Append-only store for links *)
+module type AO_LINK = sig
+
+
+  (** {1 Append-only Link stores} *)
+
+  include RO
+
+  val add: t -> key -> value -> key Lwt.t
```
Comment (samoht) [inline lib/irmin.mli]: I think I would use a different function name, as `add` is already used with a different type somewhere else. What abouht `add_link`?

File: lib/krypto/irmin_krypto.ml [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+(**
+
+  Krypto : Irmin Crypto Backend
+  The kryptonite for protect your data of nasties
+
+  TODO :
+  Padding into blobs, and cut blobs on a defined size block -> We don't want to guess the size of content
```
Comment (samoht) [inline lib/krypto/irmin_krypto.ml]: I guess that TODO is now done!

File: lib/krypto/irmin_krypto.ml [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+      
+			  
+    let add t v =
+      let ctr = compute_ctr v in
+      let ctr2 = compute_ctr v in
+      C.encrypt ~ctr v |> inject_ctr ~ctr:ctr2 |> AO.add t
+
+
+    let iter _t (_fn : key -> value Lwt.t -> unit Lwt.t) =
+      failwith "NOT NEEDED"
```
Comment (samoht) [inline lib/krypto/irmin_krypto.ml]: the tests should need this! (if not, we should fix the tests :p)

Comment (samoht) [inline lib/krypto/irmin_krypto.ml]: The tests should exercice that code-path!

**Comment by dbuenzli** [comment]
Is the `Chunck` name on purpose ? Shouldn't that be `Chunk` ?

File: lib/link/irmin_link.mli [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+    type key
+    type value	   
+    type t
+	   
+    val create: config -> 'a Task.f -> ('a -> t) Lwt.t
+    val read: t -> key -> value option Lwt.t
+    val add: t -> key -> value -> key Lwt.t
+    val iter: t -> (key -> value Lwt.t -> unit Lwt.t) -> unit Lwt.t
+    val mem: t -> key -> bool Lwt.t
+end
```
Comment (samoht) [inline lib/link/irmin_link.mli]: what's the relation between this module type and `AO_LINK`. I'm not sure to understand why do you need both.

File: lib_test/test_common.ml [hunk 26]
```diff
@@ -110,7 +110,14 @@ let create: (module Irmin.S_MAKER) -> [`String | `Json] -> (module Irmin.S) =
...
 
-type kind = [`Mem | `Fs | `Git | `Http of kind]
+type kind = [`Mem
+	    | `Fs
+	    | `Git
+	    | `Link_FS_Chunck_Krypto
+	    | `Link_MEM_Chunck_Krypto
+	    | `Link_FS_Chunck
+	    | `Link_MEM_Chunck
+	    | `Http of kind]
```
Comment (samoht) [inline lib_test/test_common.ml]: does it make sense ti test the chunk and the crypto module in isolation as well (ie. without using the link layer)?

Comment (samoht) [inline lib_test/test_common.ml]: Also I think that type would look better with something like:

```ocaml
type kind = [ 
  | `Mem
  | `Git
  | `Http of kind
  | `Krypto of kind
  | `Chunk of kind
  | `Link of kind
```

I guess all the combinations are possible although some of them don't really make sense.

File: lib_test/test_fs_link_chunck.ml [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+
+open Test_common
+
+let test_db = "test-db-chunck"
+
+let polling =
+  try float_of_string (Sys.getenv "IRMIN_FS_POLLING")
+  with Not_found | Failure _ -> 0.01
+
+let init () =
```
Comment (samoht) [inline lib_test/test_fs_link_chunck.ml]: Use `Test_fs.init` instead of duplicating code.

File: lib_test/test_fs_link_chunck.ml [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+  Irmin_unix.install_dir_polling_listener polling;
+  if Sys.file_exists test_db then begin
+    let cmd = Printf.sprintf "rm -rf %s" test_db in
+    let _ = Sys.command cmd in ()
+  end;
+  Lwt.return_unit
+
+let clean () =
```
Comment (samoht) [inline lib_test/test_fs_link_chunck.ml]: Use `Test_fs.clean` instead of duplicating code.

File: lib_test/test_fs_link_chunck_krypto.ml [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+
+open Test_common
+
+let test_db = "test-db-chunck"
+
+let polling =
+  try float_of_string (Sys.getenv "IRMIN_FS_POLLING")
+  with Not_found | Failure _ -> 0.01
+
+let init () =
```
Comment (samoht) [inline lib_test/test_fs_link_chunck_krypto.ml]: Use `Test_fs.init` and `Test_fs.clean`

File: lib_test/test_fs_link_chunck.ml [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+  Irmin_unix.uninstall_dir_polling_listener ();
+  Lwt.return_unit
+    
+let config = Irmin_fs.config ~root:test_db ()
+let config = Irmin_chunck.config ~conf:config ~size:4096 ()
+    
+let suite k =
+  {
+    name   = "LINK FS - CHUNCK" ^ string_of_contents k;
```
Comment (samoht) [inline lib_test/test_fs_link_chunck.ml]: the test name should not have spaces in it. They are used to select sub-tests on the command line so spaces are annoying there.

File: lib_test/test_fs_link_chunck.ml [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+    kind   = `Link_FS_Chunck;
+    cont   = k;
+    init   = init;
+    clean  = clean;
+    store  = link_fs_chunck_store k;
+    config = config;
+  }
```
Comment (samoht) [inline lib_test/test_fs_link_chunck.ml]: All the new tests will benefit avoiding code duplication and be more functorised. I don't really want to copy/past more code if I decide to test chunk over chunck over http over memory. I just want to apply a functor (see how it''s done for the http backend tests).

File: lib/unix/irmin_unix.mli [hunk 32]
```diff
@@ -123,6 +126,10 @@ module Irmin_git: sig
       be written in {i .git/objects/} and might be cleaned-up if you
       run {i git gc} manually. *)
 
+(*  module AO_LINK (G: Git.Store.S): Irmin.AO_LINK_MAKER
+  (** Embed an append-only store for links into a Git repository. Contents will
+      be written in {i .git/links/}. *)
+ *)
```
Comment (samoht) [inline lib/unix/irmin_unix.mli]: remove dead-code.

**Comment by samoht** [comment]
Thanks for the update, I've added few minor comments that need to be addressed before merging. To summarise:

- rename `Chunck` to `Chunk` (thanks Daniel :p)
- remove all the dead-code: commented code, non-linking files and space patches
- sort-out the duplication between `AO_LINK` and `Irmin_link`
- functorise the test code to be able to tests more complex backend combination if needed (and to avoid code duplication)
- also before merging it would be nice to clean-up the history. Ideally, I would recommend to have multiple commits:
   - one to modify `Irmin` to add the `AO_LINK` functors.
   - one to add `AO_RAW` in `Irmin`.
   - one to add the `chunk/` code
   - one to add the `krypto/` code

I'm happy to help your with the various Git commands involved to do a proper rebase, just come to see me when you are around :-)

**Comment by samoht** [comment]
Irmin API changes are now part of master, and [irmin-chunk](https://github.com/mirage/irmin-chunk] and [irmin-crypto](https://github.com/mirage/irmin-krypto) have been created as external libraries.

So closing this PR.
